pkg/controller/batchrelease/workloads/unified_workload_control_plane.go

```diff
@@ -0,0 +1,316 @@
+limitations under the License.
+*/
+
+package workloads
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/openkruise/rollouts/api/v1alpha1"
+	"github.com/openkruise/rollouts/pkg/util"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type UnifiedWorkloadController interface {
+	GetWorkloadInfo() (*util.WorkloadInfo, error)
+	ClaimWorkload() (bool, error)
+	ReleaseWorkload(cleanup bool) (bool, error)
+	UpgradeBatch(canaryReplicasGoal, stableReplicasGoal int32) (bool, error)
+	IsBatchUpgraded(canaryReplicasGoal, stableReplicasGoal int32) (bool, error)
+	IsBatchReady(canaryReplicasGoal, stableReplicasGoal int32) (bool, error)
+}
+
+// UnifiedWorkloadRolloutControlPlane is responsible for handling rollout StatefulSet type of workloads
+type UnifiedWorkloadRolloutControlPlane struct {
+	UnifiedWorkloadController
+	client         client.Client
+	recorder       record.EventRecorder
+	planController *v1alpha1.BatchRelease
+	newStatus      *v1alpha1.BatchReleaseStatus
+}
+
+type NewUnifiedControllerFunc = func(c client.Client, r record.EventRecorder, p *v1alpha1.BatchRelease, n types.NamespacedName, gvk schema.GroupVersionKind) UnifiedWorkloadController
+
+// NewUnifiedWorkloadRolloutControlPlane creates a new CloneSet rollout controller
+func NewUnifiedWorkloadRolloutControlPlane(f NewUnifiedControllerFunc, c client.Client, r record.EventRecorder, p *v1alpha1.BatchRelease, newStatus *v1alpha1.BatchReleaseStatus, n types.NamespacedName, gvk schema.GroupVersionKind) *UnifiedWorkloadRolloutControlPlane {
+	return &UnifiedWorkloadRolloutControlPlane{
+		client:                    c,
+		recorder:                  r,
+		planController:            p,
+		newStatus:                 newStatus,
+		UnifiedWorkloadController: f(c, r, p, n, gvk),
+	}
+}
+
+// VerifyWorkload verifies that the workload is ready to execute release plan
+func (c *UnifiedWorkloadRolloutControlPlane) VerifyWorkload() (bool, error) {
+	var err error
+	defer func() {
+		if err != nil {
+			klog.Warningf(err.Error())
+			c.recorder.Event(c.planController, v1.EventTypeWarning, "VerifyFailed", err.Error())
+		}
+	}()
+
+	workloadInfo, err := c.GetWorkloadInfo()
+	if err != nil {
+		return true, err
+	}
+
+	// If the workload status is untrustworthy
+	if workloadInfo.Status.ObservedGeneration != workloadInfo.Metadata.Generation {
+		err = fmt.Errorf("%v is still reconciling, wait for it to be done", workloadInfo.GVKWithName)
+		return false, err
+	}
+
+	// If the workload has been promoted, no need to go on
+	if workloadInfo.Status.UpdatedReplicas == *workloadInfo.Replicas {
+		err = fmt.Errorf("%v update revision has been promoted, no need to reconcile", workloadInfo.GVKWithName)
+		return false, err
+	}
+
+	// If the workload is not paused and is not under our control
+	if !workloadInfo.Paused {
```

Comment:
we shall also unpause statefulset in resumeStableWorkload e.g. set partition to nil